core/src/main/java/org/apache/calcite/rel/core/JoinRelType.java

```diff
@@ -152,4 +152,29 @@ public JoinRelType cancelNullsOnRight() {
   public boolean projectsRight() {
     return this != SEMI && this != ANTI;
   }
+
+  /** Returns whether this join type accepts pushing predicates from above into its predicate. */
+  public boolean canPushIntoFromAbove() {
+    return (this == INNER) || (this == SEMI);
```

Comment:
Are we sure ANTI does not belong to `canPushIntoFromAbove`?

Reply:
I'm thinking about this in terms of the algebra rather than SQL. The rule can match a filter on top of ANTI, and you are trying to push those expressions from above into the join condition, to the left input as a filter, or the right input as a filter. `canPushLeftFromAbove` returns true for ANTI indeed as you described. However, `canPushIntoFromAbove` returns false for ANTI because it's not semantically equivalent to filter after ANTI and to filter within the ANTI. SEMI semantics are different.